drivers/ethernet/eth_stm32_hal.c

```diff
@@ -227,10 +227,17 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
 	/* Reset TX complete interrupt semaphore before TX request*/
 	k_sem_reset(&dev_data->tx_int_sem);
 
+	/* Memory barriers (e.g. __DSB()) should flush Tx descriptors
+	 * into memory. Incomplete writing of Tx descriptors may occur
+	 * with different compilation levels
+	 */
+	__DSB();
+
 	/* tx_buffer is allocated on function stack, we need */
 	/* to wait for the transfer to complete */
 	/* So it is not freed before the interrupt happens */
 	hal_ret = HAL_ETH_Transmit_IT(heth, &tx_config);
+	__DSB();
 
```

Comment:
```suggestion
__ISB();
```

Can you please put `__ISB()` here and try last test wit `CONFIG_NO_OPTIMIZATIONS=y`? According to [2.3.4
Software ordering of memory accesses](https://www.st.com/resource/en/programming_manual/dm00237416-stm32f7-series-and-stm32h7-series-cortexm7-processor-programming-manual-stmicroelectronics.pdf):

**DSB**
The Data Synchronization Barrier (DSB) instruction ensures that outstanding memory transactions complete before subsequent
instructions execute. See DSB on page 177.

**ISB**
The Instruction Synchronization Barrier (ISB) ensures that the effect of all completed memory transactions is recognizable by subsequent
instructions. See ISB on page 178.

Reply:
I tried this and I still get the semaphore timeout.I've tried all combinations of ordering and presence.